### ISSUE TITLE
[Proposal] #await (inspired by Svelte)

### ISSUE THREAD

**Issue by dmitriid** [surface-ui/surface#496]
## What

This is inspired by Svelte's [#await](https://svelte.dev/docs#await). While Phoenix doesn't have promises, we can use `#await` as a QoL improvement primarily when dealing with `nil` values in assigns.

Feel free to laugh and close :)

## Quick overview (but do read on to rationale and syntax)

<table><thead><tr><th>Proposed</th><th>Equivalent to</th></tr></thead>
  <tbody>
  <tr>
    <td>
      <pre>
{#await @value }
  Loading...
{#then}
  {@value}
{#catch}
  An error occurred
{/await}
      </pre>
    </td>
    <td>
      <pre>
{#case assigns[:value]}
  {#match nil}
    Loading...
  {#match {:error, _}}
    An error occured
  {#match _}
    {@value}
{/case}
      </pre>
    </td>
  </tr>
  <tr><td>This can be mixed and matched</td><td></td></tr>
  <tr>
    <td>
      <pre>
{#await @value }
  Loading...
{/await}
      </pre>
    </td>
    <td>
      <pre>
{#if assigns[:value] == nil }
  Loading...
{/if}
      </pre>
    </td>
  </tr>
  <tr>
    <td>
      <pre>
{#await @value }
  Loading...
{#then}
  {@value}
{/await}
      </pre>
    </td>
    <td>
      <pre>
{#case assigns[:value]}
  {#match nil}
    Loading...
  {#match {:error, _}}
  {#match _}
    {@value}
{/case}
      </pre>
    </td>
  </tr>
  <tr>
    <td>
      <pre>
{#await @value }
  Loading...
{#catch}
  An error occured
{/await}
      </pre>
    </td>
    <td>
      <pre>
{#case assigns[:value]}
  {#match nil}
    Loading...
  {#match {:error, _}}
    An error occured
  {#match _}
{/case}
      </pre>
    </td>
  </tr>
  </tbody>
</table>

## Rationale

A UI state is a fluid thing. Often, as a shortcut, you would want to just skip setting a value until it's needed. Unfortunately, dealing with such a case is a bit cumbersome:

This code will break with `assign @non_existent_value not available in eex template.`
```
{ @non_existent_value } 
```

To mitigate this, the code has to be written like this:

```
{#if assigns[:non_existent_value] != nil }{ @non_existent_value }{/if}
```

This becomes somewhat weird when existence/non-existence is key:

```
{#if assigns[:value] == nil}
  do one thing
{#else}
  do another thing with {@value}
{/if}
```

`#await` allows you to deal with these and other situations easily, and be very explicit about what it is you want from that value.

## Syntax

Directly lifted from Svelte:

```
{#await expression}...{#then name}...{#catch name}...{/await}
{#await expression}...{#then name}...{/await}
{#await expression then name}...{/await}
{#await expression catch name}...{/await}
```

### If we allow any `expression`, taking care of assigns once again becomes cumbersome or magical

```
{#await @x == {:ok, [@y]} then}...{/await}
```

Do we replace `@x` and `@y` here with `assigns[:x]` and `assigns[:y]`?

### Adopting `{#then var_name}` and `{#catch var_name}`

This becomes repetitive for matching against assigns:

```
{#await @x then x} { x } {/await}
```

But works well with expressions:

```
{#await @x == {:ok, [@y]} then z} {z} {/await}
```

If `#await# is expanded into a case, this becomes similar to regular pattern matching, so not really a problem, and you could probably have several catch cases for errors:

```
{#await expression}
  {#catch {:error, reason}}
     Error: {reason}
  {#catch {:error, other_reason}}
     Error: {other_reason}
{/await}
```

**Comment by paulstatezny** [comment]
I don't have strong opinions about this proposal. However, if this proposal moves forward we'll want to make sure the mechanism doesn't break change tracking.

There are some [LiveView pitfalls](https://hexdocs.pm/phoenix_live_view/assigns-eex.html#pitfalls) whereby the user can break change tracking, causing lots of unchanged HTML to be sent over the wire every time any event happens. Referencing `assigns` directly causes this, and I see this proposal is built on performing `assigns[:some_assign_that_may_or_may_not_exist]`. I'm not sure if Surface could do this check without breaking change tracking.

**Comment by davydog187** [comment]
Implementation aside, the naming of any DSL here probably needs some thought. In Javascript, `await` makes sense as it aligns with `async/await` and the `Promises` API. Elixir does not have those concepts so it seems a bit off

**Comment by paulstatezny** [comment]
@davydog187 I agree, [catch](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#try/1) and [await](https://hexdocs.pm/elixir/Task.html#await/2) are already in the Elixir standard lib, with meanings that are unrelated to checking for `{:error, _}` tuples and `nil`.

But the naming isn't my main concern. The current special tags in Surface (`if`, `for`, `case`) follow their Elixir analogues very closely. Elixir devs should be able to grasp them with relative ease. However, this proposal doesn't have an Elixir analogue. That's not inherently bad, but I'm questioning the value proposition. What do we get in return for adding more syntax "surface area"? (No pun intended 😁)

Perhaps I'm missing something, but it appears the benefit is slightly more succinct code. Devs can achieve the same thing with `case`/`if` structures, as the examples above demonstrate. `case` and `if` are more explicit about what Boolean logic and pattern matching is actually happening, rather than hiding those details behind a keyword. If this proposal cut down the amount of code by a significant portion, I'd be more apt to support it. But I think the existing structures are already relatively succinct.

In summary, I think Surface should be designed for clarity in its context (Elixir) rather than optimized for maximally succinct code.

**Comment by dmitriid** [comment]
Yes, all your concerns are very valid. It's too big of a change for a minor quality-of-life improvement

My own wish for a construct like this comes simply from the awkwardness of dealing with nil values especially in assigns.

I'll have to think more on this :)

**Comment by paulstatezny** [comment]
That makes sense @dmitriid. Don't mean to violently shut you down with my response 😅 I really appreciate the time and thought you obviously put into the proposal!
